Plantable Quartz

Dariensg opened the pull request:
-Added Plantable Quartz
-Added Plantable Quartz Feature
-Apparently fixed Wildcard import in IEBlocks (You're welcome Swan)

Dariensg committed:
Plantable Quartz

-Added Plantable Quartz
-Added Plantable Quartz Feature
-Apparently fixed Wildcard import in IEBlocks (You're welcome Swan)

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/blocks/PlantedQuartzBlock.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    @Override
+    protected boolean isValidGround(BlockState state, IBlockReader worldIn, BlockPos pos) {
+        return state.matchesBlock(Blocks.QUARTZ_BLOCK) || state.matchesBlock(Blocks.CHISELED_QUARTZ_BLOCK)
+            || state.matchesBlock(Blocks.SMOOTH_QUARTZ) || state.matchesBlock(Blocks.QUARTZ_BRICKS)
+            || state.matchesBlock(Blocks.QUARTZ_PILLAR) || state.matchesBlock(Blocks.NETHER_QUARTZ_ORE)
+            || state.matchesBlock(Blocks.SOUL_SAND) || state.matchesBlock(Blocks.SOUL_SOIL)
+            || state.matchesBlock(Blocks.NETHERRACK);
+    }
```
convert this to a tag

Hellion-Games replied:
This is what the BuriedBoneBlock does. Should I change that one too?

SwanX1 replied:
yes

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/events/MiscEvents.java:
```diff
@@ -129,6 +129,23 @@ public void onRightClickBlock(PlayerInteractEvent.RightClickBlock event) {
                 world.setBlockState(pos, IEBlocks.DIMSTONE.get().getDefaultState());
             }
         }
+
+        if (heldItemStack.getItem() == Items.QUARTZ) {
+            pos = pos.offset(face);
+            BlockState blockstate = IEBlocks.PLANTED_QUARTZ.get().getPlaceableState(world, pos, face);
+            if (blockstate != null) {
+                player.swingArm(event.getHand());
+                if (!world.isAirBlock(pos) && !world.isRemote() && world.getBlockState(pos).getFluidState().isEmpty()) {
+                    world.destroyBlock(pos, true);
+                }
+                world.setBlockState(pos, blockstate, 3);
+                world.playSound(player, pos, blockstate.getSoundType().getPlaceSound(), SoundCategory.BLOCKS, 1.0F, 1.0F);
+                if (!player.isCreative()) {
+                    heldItemStack.shrink(1);
+                }
+                ForgeEventFactory.onBlockPlace(player, BlockSnapshot.create(world.getDimensionKey(), world, pos), face);
+            }
+        }
```
if you could move this right below the bone if statement that would be epic (using an `else if`)

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/world/gen/features/PlantedQuartzFeature.java:
```diff
@@ -0,0 +1,98 @@
+
+        int pickFacing = random.nextInt(4);
+
+        if (pickFacing <= 1) {
+            face = AttachFace.FLOOR;
+        } else if (pickFacing == 2) {
+            face = AttachFace.WALL;
+            int pickDirection = random.nextInt(4);
+
+            if (pickDirection == 0) {
+                direction = Direction.NORTH;
+            } else if (pickDirection == 1) {
+                direction = Direction.SOUTH;
+            } else if (pickDirection == 2) {
+                direction = Direction.EAST;
+            } else {
+                direction = Direction.WEST;
+            }
+        } else {
+            face = AttachFace.CEILING;
+        }
```
I think it would be better if you checked for the placeable faces and then chose from the available faces instead of choosing a random face and hoping that it would generate on a vaild face.

Hellion-Games replied:
How would you propose is the best way to make that check?

SwanX1 replied:
*responded in discord*

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/world/gen/features/PlantedQuartzFeature.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    public boolean findOre(ISeedReader world, BlockPos pos) {
+        boolean foundOre = false;
+        final int radius = 3;
+        outerLoop:
+        for (int x = pos.getX() - radius; x < pos.getX() + radius; x++) {
+            for (int y = pos.getY() - radius; y < pos.getY() + radius; y++) {
+                for (int z = pos.getZ() - radius; z < pos.getZ() + radius; z++) {
+                    if (world.getBlockState(new BlockPos(x, y, z)).getBlock().equals(Blocks.NETHER_QUARTZ_ORE)) {
+                        foundOre = true;
+                        break outerLoop;
+                    }
+                }
+            }
+        }
+        return foundOre;
```
```suggestion
        final int radius = 3;
        for (int x = pos.getX() - radius; x < pos.getX() + radius; x++) {
            for (int y = pos.getY() - radius; y < pos.getY() + radius; y++) {
                for (int z = pos.getZ() - radius; z < pos.getZ() + radius; z++) {
                    if (world.getBlockState(new BlockPos(x, y, z)).getBlock().equals(Blocks.NETHER_QUARTZ_ORE)) {
                        return true;
                    }
                }
            }
        }
        return false;
```

SwanX1 reviewed (changes requested):
A block translation key would be good for tooltip mods like TOP and HWYLA.

Dariensg committed:
Update src/main/java/com/nekomaster1000/infernalexp/world/gen/features/PlantedQuartzFeature.java

Co-authored-by: Kārlis Čerņavskis <karliscern@gmail.com>

Dariensg committed:
PR Suggestion Fixes

-Moved isValidGround blocks into a tag for quartz and bone
-Made the Feature generation choose between valid faces (credit to Swan)
-Added translation key for Planted Quartz
-Refactored MiscEvents
-Updated inaccurate feature comment
-Updated Lang Files

SwanX1 reviewed (approved):
LGTM, merging.
